### ISSUE TITLE
[$500 Bounty]☂️  Training workload for tiiuae/Falcon3-1B-Base

### ISSUE THREAD

**Issue by vladimirjovanovicTT** [tenstorrent/tt-blacksmith#212]
## Add Falcon3-1B-Base Single-Chip LoRA Training on TT-N150 with CPU Baseline  

**Background:**  
- TT-Blacksmith is a growing collection of model demos showcasing the capabilities of AI models running on Tenstorrent hardware.  
- This bounty aims to bring up **full LoRA fine-tuning** of **tiiuae/Falcon3-1B-Base** on a single **Tenstorrent N150** chip, faithfully reproducing the training setup in PyTorch.  
- To establish a baseline, the same training must also be run on CPU and the results compared, ensuring that metrics (e.g., loss curves, validation accuracy, final perplexity) align between CPU and TT-N150 runs.  

When encountering compilation or runtime issues, contributors are expected to:  
1. Delegate the issue with a minimal repro to the appropriate Tenstorrent repository (```tt-xla/tt-mlir``` for compilation, ```tt-metal``` for runtime).  
2. Provide a **CPU fallback workaround** in their training code so training can continue while the issue is investigated.  

This bounty is an opportunity for AI developers to:  
- Contribute to Tenstorrent’s model demos  
- Validate single-chip LoRA training  
- Earn rewards  

---

**Requirements:**  
- **Framework:** Implementation must be in **PyTorch**.  

1. **Falcon3-1B LoRA Training on TT-N150:**  
   - Implement **end-to-end LoRA fine-tuning** on a single chip.  
   - Fine-tune the model on the **Wikitext-2** dataset.  
   - Reproduce all training hyperparameters, data setup, and workflow from the Falcon-1B base model.  

2. **CPU Baseline:**  
   - Run the same LoRA fine-tuning on CPU.  
   - Compare results, ensuring parity in metrics and convergence.  

3. **Fallback Mechanism:**  
   - In cases where an operation fails on TT-N150, redirect only the **minimal possible function containing that operation** to CPU.  
   - Do **not** fallback the entire training step to CPU — training must primarily execute on TT hardware.  

4. **Verification:**  
   - To confirm training is running on TT hardware, set:  
     ```bash
     export LOGGER_LEVEL=DEBUG
     ```  
     - Look for printed TTIR graphs (e.g., `graph module @jit_training_step`) confirming TT execution.  

5. **Koyeb Machines:**  
   - Training should be run on **Koyeb-hosted N150 instances**.  
   - Request access here: [https://www.koyeb.com/solutions/tenstorrent](https://www.koyeb.com/solutions/tenstorrent)  
   - Wait for the onboarding email from Koyeb with setup instructions.  

6. **Documentation:**  
   - Provide a detailed **README.md** including:  
     - Setup instructions for both CPU and TT-N150  
     - Explanation of LoRA setup and configuration  
     - Explanation of fallback logic  

7. **Examples:**  
   - Include sample input/output and **plots of loss/accuracy curves** for CPU vs TT-N150 runs.  
   - Include example LoRA configuration and checkpoints.  

8. **Dependencies:**  
   - Clearly document installation steps and dependencies.  

---

**Contribution Guidelines:**  
1. Fork the **tt-blacksmith** repository.  
2. Create a directory:  
   ```
   tt-blacksmith/blacksmith/experiments/torch/BOUNTIES/falcon3-1b
   ```  
3. Add your model implementation, training scripts, and documentation there.  
4. Submit a pull request with:  
   - A detailed description  
   - Results and comparison plots  
   - Relevant information to help reviewers evaluate your contribution  

---

**Evaluation Criteria:**  
1. **Framework Compliance:** Implementation must be in PyTorch.  
2. **Hardware Requirement:** Must demonstrate training on a single TT-N150 chip (via Koyeb).  
3. **Completeness:** End-to-end LoRA fine-tuning and training must be functional.  
4. **Metric Parity:** CPU vs TT-N150 runs should show comparable convergence (loss, perplexity, etc.).  
5. **Fallback Implementation:** Only targeted CPU fallbacks are acceptable — the full training step must not run on CPU.  
6. **Verification:** Use `LOGGER_LEVEL=DEBUG` to confirm TTIR execution of key training functions.  
7. **Minimal Repros:** Any delegated issue must include a minimal reproducer.  
8. **Code Quality:** Clear, maintainable, and well-commented code.  
9. **Documentation:** Setup and usage must be clearly explained, including LoRA setup.

**Comment by rishi-jat** [comment]
@vladimirjovanovicTT can I work on this issue as I see this bounty is open ?

**Comment by samiyadev786** [comment]
@marty1885 Please assign me?

**Comment by marty1885** [comment]
@samiyadev786 Can you share some progress you have made?
